# Arkadiusz Pytlik

In [1]:
import psycopg2 as pg
import pandas as pd

connection = pg.connect(host='pgsql-196447.vipserv.org', port=5432, dbname='wbauer_adb', user='wbauer_adb', password='adb2020')

### 1. Ile kategorii filmów mamy w wypożyczalni?

In [2]:
df = pd.read_sql('SELECT COUNT (DISTINCT name) FROM category',con=connection)
df

count
0     16

W wypożyczalni jest 16 kategorii filmów.

### 2. Wyświetl listę kategorii w kolejności alfabetycznej.

In [3]:
df = pd.read_sql('SELECT name FROM category ORDER BY name ASC',con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

### 3. Znajdź najstarszy i najmłodszy film do wypożyczenia.

In [4]:
df = pd.read_sql('SELECT title, release_year FROM film ORDER BY release_year ASC LIMIT 1',con=connection)
df

title  release_year
0  Chamber Italian          2006

In [5]:
df = pd.read_sql('SELECT title, release_year FROM film ORDER BY release_year DESC LIMIT 1',con=connection)
df

title  release_year
0  Chamber Italian          2006

In [6]:
df = pd.read_sql('SELECT DISTINCT release_year FROM film',con=connection)
df


release_year
0          2006

Wyszystkie filmy są z tego samego roku, więc zapytania zwracają te same filmy.

### 4. Ile wypożyczeń odbyło się między 2005-07-01 a 2005-08-01?

In [7]:
df = pd.read_sql("SELECT COUNT(*) FROM rental WHERE rental_date BETWEEN '2005-07-01' AND '2005-08-01'",con=connection)
df

count
0   6709

W tym przedziale czasowym odbyło się 6709 wypożyczeń.

### 5. Ile wypożyczeń odbyło się między 2010-01-01 a 2011-02-01?

In [8]:
df = pd.read_sql("SELECT COUNT(*) FROM rental WHERE rental_date BETWEEN '2010-01-01' AND '2011-02-01'",con=connection)
df

count
0      0

W tym przedziale czasowym nic nie wypożyczono.

### 6. Znajdź największą płatność wypożyczenia.

In [9]:
df = pd.read_sql('SELECT amount FROM payment ORDER BY amount DESC LIMIT 1',con=connection)
df

amount
0   11.99

Największa płatność wynosiła 11.99.

### 7. Znajdź wszystkich klientów z Polski, Nigerii lub Bangladeszu.

In [10]:
df = pd.read_sql("SELECT country_id FROM country WHERE country IN ('Poland', 'Nigeria', 'Bangladesh')",con=connection)
country_id = tuple(df['country_id'].tolist())

df = pd.read_sql(f"SELECT city_id FROM city WHERE country_id IN {country_id}",con=connection)
city_id = tuple(df['city_id'].tolist())

df = pd.read_sql(f"SELECT address_id FROM address WHERE city_id IN {city_id}",con=connection)
address_id = tuple(df['address_id'].tolist())

df = pd.read_sql(f"SELECT * FROM customer WHERE address_id IN {address_id}",con=connection)
df

customer_id  store_id first_name  last_name  \
0            18         2      Carol     Garcia   
1            21         1   Michelle      Clark   
2            80         1    Marilyn       Ross   
3           103         1     Gladys   Hamilton   
4           128         1   Marjorie     Tucker   
5           156         1     Bertha   Ferguson   
6           198         2      Elsie     Kelley   
7           232         2  Constance       Reid   
8           250         2         Jo     Fowler   
9           270         1       Leah     Curtis   
10          277         2       Olga    Jimenez   
11          280         2     Tracey    Barrett   
12          284         1      Sonia    Gregory   
13          286         1      Velma      Lucas   
14          318         1      Brian      Wyman   
15          329         2      Frank   Waggoner   
16          332         1    Stephen     Qualls   
17          380         1    Russell    Brinson   
18          409         2     Rodney    Moeller   
19          501         1      Ruben      Geary   
20          558         1     Jimmie  Eggleston   
21          562         1    Wallace      Slone   
22          571         2    Johnnie   Chisholm   
23          572         1     Sidney   Burleson   

                                  email  address_id  activebool create_date  \
0       carol.garcia@sakilacustomer.org          22        True  2006-02-14   
1     michelle.clark@sakilacustomer.org          25        True  2006-02-14   
2       marilyn.ross@sakilacustomer.org          84        True  2006-02-14   
3    gladys.hamilton@sakilacustomer.org         107        True  2006-02-14   
4    marjorie.tucker@sakilacustomer.org         132        True  2006-02-14   
5    bertha.ferguson@sakilacustomer.org         160        True  2006-02-14   
6       elsie.kelley@sakilacustomer.org         202        True  2006-02-14   
7     constance.reid@sakilacustomer.org         236        True  2006-02-14   
8          jo.fowler@sakilacustomer.org         254        True  2006-02-14   
9        leah.curtis@sakilacustomer.org         275        True  2006-02-14   
10      olga.jimenez@sakilacustomer.org         282        True  2006-02-14   
11    tracey.barrett@sakilacustomer.org         285        True  2006-02-14   
12     sonia.gregory@sakilacustomer.org         289        True  2006-02-14   
13       velma.lucas@sakilacustomer.org         291        True  2006-02-14   
14       brian.wyman@sakilacustomer.org         323        True  2006-02-14   
15    frank.waggoner@sakilacustomer.org         334        True  2006-02-14   
16    stephen.qualls@sakilacustomer.org         337        True  2006-02-14   
17   russell.brinson@sakilacustomer.org         385        True  2006-02-14   
18    rodney.moeller@sakilacustomer.org         414        True  2006-02-14   
19       ruben.geary@sakilacustomer.org         506        True  2006-02-14   
20  jimmie.eggleston@sakilacustomer.org         564        True  2006-02-14   
21     wallace.slone@sakilacustomer.org         568        True  2006-02-14   
22  johnnie.chisholm@sakilacustomer.org         577        True  2006-02-14   
23   sidney.burleson@sakilacustomer.org         578        True  2006-02-14   

               last_update  active  
0  2013-05-26 14:49:45.738       1  
1  2013-05-26 14:49:45.738       1  
2  2013-05-26 14:49:45.738       1  
3  2013-05-26 14:49:45.738       1  
4  2013-05-26 14:49:45.738       1  
5  2013-05-26 14:49:45.738       1  
6  2013-05-26 14:49:45.738       1  
7  2013-05-26 14:49:45.738       1  
8  2013-05-26 14:49:45.738       1  
9  2013-05-26 14:49:45.738       1  
10 2013-05-26 14:49:45.738       1  
11 2013-05-26 14:49:45.738       1  
12 2013-05-26 14:49:45.738       1  
13 2013-05-26 14:49:45.738       1  
14 2013-05-26 14:49:45.738       1  
15 2013-05-26 14:49:45.738       1  
16 2013-05-26 14:49:45.738       1  
17 2013-05-26 14:49:45.738       1  
18 2013-05-26 14:49:45.738       1  
19 2013-05-2

In [11]:
# Alternatywna metoda
df = pd.read_sql("SELECT customer.* FROM country " +
                 "INNER JOIN city ON country.country_id = city.country_id " +
                 "INNER JOIN address ON city.city_id = address.city_id " +
                 "INNER JOIN customer ON address.address_id = customer.address_id " +
                 "WHERE country.country IN ('Poland', 'Nigeria', 'Bangladesh') " +
                 "ORDER BY customer_id ASC",con=connection)
df

customer_id  store_id first_name  last_name  \
0            18         2      Carol     Garcia   
1            21         1   Michelle      Clark   
2            80         1    Marilyn       Ross   
3           103         1     Gladys   Hamilton   
4           128         1   Marjorie     Tucker   
5           156         1     Bertha   Ferguson   
6           198         2      Elsie     Kelley   
7           232         2  Constance       Reid   
8           250         2         Jo     Fowler   
9           270         1       Leah     Curtis   
10          277         2       Olga    Jimenez   
11          280         2     Tracey    Barrett   
12          284         1      Sonia    Gregory   
13          286         1      Velma      Lucas   
14          318         1      Brian      Wyman   
15          329         2      Frank   Waggoner   
16          332         1    Stephen     Qualls   
17          380         1    Russell    Brinson   
18          409         2     Rodney    Moeller   
19          501         1      Ruben      Geary   
20          558         1     Jimmie  Eggleston   
21          562         1    Wallace      Slone   
22          571         2    Johnnie   Chisholm   
23          572         1     Sidney   Burleson   

                                  email  address_id  activebool create_date  \
0       carol.garcia@sakilacustomer.org          22        True  2006-02-14   
1     michelle.clark@sakilacustomer.org          25        True  2006-02-14   
2       marilyn.ross@sakilacustomer.org          84        True  2006-02-14   
3    gladys.hamilton@sakilacustomer.org         107        True  2006-02-14   
4    marjorie.tucker@sakilacustomer.org         132        True  2006-02-14   
5    bertha.ferguson@sakilacustomer.org         160        True  2006-02-14   
6       elsie.kelley@sakilacustomer.org         202        True  2006-02-14   
7     constance.reid@sakilacustomer.org         236        True  2006-02-14   
8          jo.fowler@sakilacustomer.org         254        True  2006-02-14   
9        leah.curtis@sakilacustomer.org         275        True  2006-02-14   
10      olga.jimenez@sakilacustomer.org         282        True  2006-02-14   
11    tracey.barrett@sakilacustomer.org         285        True  2006-02-14   
12     sonia.gregory@sakilacustomer.org         289        True  2006-02-14   
13       velma.lucas@sakilacustomer.org         291        True  2006-02-14   
14       brian.wyman@sakilacustomer.org         323        True  2006-02-14   
15    frank.waggoner@sakilacustomer.org         334        True  2006-02-14   
16    stephen.qualls@sakilacustomer.org         337        True  2006-02-14   
17   russell.brinson@sakilacustomer.org         385        True  2006-02-14   
18    rodney.moeller@sakilacustomer.org         414        True  2006-02-14   
19       ruben.geary@sakilacustomer.org         506        True  2006-02-14   
20  jimmie.eggleston@sakilacustomer.org         564        True  2006-02-14   
21     wallace.slone@sakilacustomer.org         568        True  2006-02-14   
22  johnnie.chisholm@sakilacustomer.org         577        True  2006-02-14   
23   sidney.burleson@sakilacustomer.org         578        True  2006-02-14   

               last_update  active  
0  2013-05-26 14:49:45.738       1  
1  2013-05-26 14:49:45.738       1  
2  2013-05-26 14:49:45.738       1  
3  2013-05-26 14:49:45.738       1  
4  2013-05-26 14:49:45.738       1  
5  2013-05-26 14:49:45.738       1  
6  2013-05-26 14:49:45.738       1  
7  2013-05-26 14:49:45.738       1  
8  2013-05-26 14:49:45.738       1  
9  2013-05-26 14:49:45.738       1  
10 2013-05-26 14:49:45.738       1  
11 2013-05-26 14:49:45.738       1  
12 2013-05-26 14:49:45.738       1  
13 2013-05-26 14:49:45.738       1  
14 2013-05-26 14:49:45.738       1  
15 2013-05-26 14:49:45.738       1  
16 2013-05-26 14:49:45.738       1  
17 2013-05-26 14:49:45.738       1  
18 2013-05-26 14:49:45.738       1  
19 2013-05-2

Obie metody zwracają te same rezultaty.

### 8. Gdzie mieszkają członkowie personelu?

In [12]:
df = pd.read_sql("SELECT first_name, last_name, address.address FROM staff " +
                 "LEFT JOIN address ON staff.address_id = address.address_id",con=connection)
df

first_name last_name               address
0       Mike   Hillyer     23 Workhaven Lane
1        Jon  Stephens  1411 Lillydale Drive

### 9. Ilu pracowników mieszka w Argentynie lub Hiszpanii?

In [13]:
df = pd.read_sql("SELECT country_id FROM country WHERE country IN ('Argentina', 'Spain')",con=connection)
country_id = tuple(df['country_id'].tolist())

df = pd.read_sql(f"SELECT city_id FROM city WHERE country_id IN {country_id}",con=connection)
city_id = tuple(df['city_id'].tolist())

df = pd.read_sql(f"SELECT address_id FROM address WHERE city_id IN {city_id}",con=connection)
address_id = tuple(df['address_id'].tolist())

df = pd.read_sql(f"SELECT COUNT(*) FROM staff WHERE address_id IN {address_id}",con=connection)
df

count
0      0

In [14]:
# Alternatywna metoda
df = pd.read_sql("SELECT COUNT(*) FROM country " +
                 "INNER JOIN city ON country.country_id = city.country_id " +
                 "INNER JOIN address ON city.city_id = address.city_id " +
                 "INNER JOIN staff ON address.address_id = staff.address_id " +
                 "WHERE country.country IN ('Argentina', 'Spain')" ,con=connection)
df

count
0      0

Żaden z pracowników nie pracuje w Argentynie ani Hiszpanii. Obie metody dają te same rezultaty.

### 10. Jakie kategorie filmów zostały wypożyczone przez klientów?

In [15]:
df = pd.read_sql("SELECT DISTINCT inventory_id FROM rental",con=connection)
inventory_id = tuple(df['inventory_id'].tolist())

df = pd.read_sql(f"SELECT DISTINCT film_id FROM inventory WHERE inventory_id IN {inventory_id}",con=connection)
film_id = tuple(df['film_id'].tolist())

df = pd.read_sql(f"SELECT DISTINCT category_id FROM film_category WHERE film_id IN {film_id}",con=connection)
category_id = tuple(df['category_id'].tolist())

df = pd.read_sql(f"SELECT name FROM category WHERE category_id IN {category_id}",con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

In [16]:
# Alternatywna metoda
df = pd.read_sql("SELECT DISTINCT category.name FROM category " +
                 "INNER JOIN film_category ON category.category_id = film_category.category_id " +
                 "INNER JOIN inventory ON film_category.film_id = inventory.film_id " +
                 "INNER JOIN rental ON inventory.inventory_id = rental.inventory_id " +
                 "ORDER BY name ASC",con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

Klienci wypożyczyli filmy z każdej kategorii. Obie metody dają te same rezultaty.

### 11. Znajdź wszystkie kategorie filmów wypożyczonych w Ameryce.

In [17]:
df = pd.read_sql("SELECT country_id FROM country WHERE country = 'United States'",con=connection)
country_id = tuple(df['country_id'].tolist())

df = pd.read_sql(f"SELECT city_id FROM city WHERE country_id = {country_id[0]}",con=connection)
city_id = tuple(df['city_id'].tolist())

df = pd.read_sql(f"SELECT address_id FROM address WHERE city_id IN {city_id}",con=connection)
address_id = tuple(df['address_id'].tolist())

df = pd.read_sql(f"SELECT customer_id FROM customer WHERE address_id IN {address_id}",con=connection)
customer_id = tuple(df['customer_id'].tolist())

df = pd.read_sql(f"SELECT DISTINCT inventory_id FROM rental WHERE customer_id IN {customer_id}",con=connection)
inventory_id = tuple(df['inventory_id'].tolist())

df = pd.read_sql(f"SELECT DISTINCT film_id FROM inventory WHERE inventory_id IN {inventory_id}",con=connection)
film_id = tuple(df['film_id'].tolist())

df = pd.read_sql(f"SELECT DISTINCT category_id FROM film_category WHERE film_id IN {film_id}",con=connection)
category_id = tuple(df['category_id'].tolist())

df = pd.read_sql(f"SELECT name FROM category WHERE category_id IN {category_id}",con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

In [18]:
# Alternatywna metoda
df = pd.read_sql("SELECT DISTINCT category.name FROM category " +
                 "INNER JOIN film_category ON category.category_id = film_category.category_id " +
                 "INNER JOIN inventory ON film_category.film_id = inventory.film_id " +
                 "INNER JOIN rental ON inventory.inventory_id = rental.inventory_id " +
                 "INNER JOIN customer ON rental.customer_id = customer.customer_id " +
                 "INNER JOIN address ON customer.address_id = address.address_id " +
                 "INNER JOIN city ON address.city_id = city.city_id " +
                 "INNER JOIN country ON city.country_id = country.country_id " +
                 "WHERE country.country = 'United States'"
                 "ORDER BY name ASC",con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

W klienci z USA wypożyczyli filmy z każdej kategorii. Obie metody dają te same rezultaty.

### 12. Znajdź wszystkie tytuły filmów, w których grał: Olympia Pfeiffer lub Julia Zellweger lub Ellen Presley

In [19]:
df = pd.read_sql(f"SELECT actor_id FROM actor WHERE " +
                 "(first_name = 'Olympia' AND last_name = 'Pfeiffer') OR " +
                 "(first_name = 'Julia' AND last_name = 'Zellweger') OR " +
                 "(first_name = 'Ellen' AND last_name = 'Presley')",con=connection)
actor_id = tuple(df['actor_id'].tolist())

df = pd.read_sql(f"SELECT DISTINCT film_id FROM film_actor WHERE actor_id IN {actor_id}",con=connection)
film_id = tuple(df['film_id'].tolist())

df = pd.read_sql(f"SELECT title FROM film WHERE film_id IN {film_id}",con=connection)
df

title
0            Badman Dawn
1        Bilko Anonymous
2   Breakfast Goldfinger
3      Caribbean Liberty
4       Casper Dragonfly
..                   ...
64             Turn Star
65            Wait Cider
66          Women Dorado
67             Won Dares
68         Wyoming Storm

[69 rows x 1 columns]

In [20]:
# Alternatywna metoda
df = pd.read_sql("SELECT film.title FROM film " +
                 "INNER JOIN film_actor ON film.film_id = film_actor.film_id " +
                 "INNER JOIN actor ON film_actor.actor_id = actor.actor_id " +
                 "WHERE " + 
                 "(first_name = 'Olympia' AND last_name = 'Pfeiffer') OR " +
                 "(first_name = 'Julia' AND last_name = 'Zellweger') OR " +
                 "(first_name = 'Ellen' AND last_name = 'Presley')"
                 "ORDER BY title ASC",con=connection)
df

title
0            Badman Dawn
1        Bilko Anonymous
2   Breakfast Goldfinger
3      Caribbean Liberty
4       Casper Dragonfly
..                   ...
64             Turn Star
65            Wait Cider
66          Women Dorado
67             Won Dares
68         Wyoming Storm

[69 rows x 1 columns]

Obie metody dały te same rezultaty.